nixos/loolwsd: init - LibreOffice/Collabora Online

mmilata opened the pull request:
This PR adds LibreOffice Online service which can be integrated into Nextcloud for collaborative document editing.

It also adds [Collabora Online CODE](https://www.collaboraoffice.com/code/) service which is a LibreOffice Online variant that is intended to be more stable and has available commercial support. Originally it was not my intention to package both but it's what I ended up with during the quest for reliable service, even got permission from Collabora to package it. **Question for reviewers: should NixOS ship both packages or just the upstream one (LibreOffice Online)?** The downside to adding Collabora Online is that it adds third LibreOffice variant to eat Hydra compute resources.

There are three parts:
* `libreoffice-online` package that provides websocket daemon interfacing with LibreOffice API, and javascript frontend that can talk to the daemon
* NixOS module for running the daemon with optional nginx reverse proxy
* (optional) `collabora-office` package, a third variant of LibreOffice, and `collabora-online` which is the `libreoffice-online` package built from different git branch and against the `collabora-office` package

[Rendered part of NixOS manual describing setup with NextCloud](https://lool.b42.cz/nixos-manual/index.html#module-services-nextcloud-office-suite)

###### Motivation for this change

Self-hosted collaborative document editing on NixOS.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is  > b up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### TODO
- [x] [crash on Insert > Image](https://bugs.documentfoundation.org/show_bug.cgi?id=129894)
- [x] [crash on Download As PDF](https://bugs.documentfoundation.org/show_bug.cgi?id=129895)
- [x] add setup instructions to Nextcloud module documentation
  - document that *Insert > Image* does not work with self-signed certificate
- [x] font file build grep hack
- [x] speed up chroot initialization
  - `LO_PATH` cannot be hardlinked because it points to /nix/store (cross-device, different owner), also it's set at compile time
  - try to minimize the closure copied to `/var/lib/lool/systemplate`
- [x] enable systemd sandboxing options
- [x] set `DICPATH` to point to hunspell dictionaries
- [ ] build underlying libreoffice-core with unnecessary features disabled (e.g. gui)
- [ ] tests are disabled
- [x] check with Collabora wrt trademark compliance
  > From: Jan Holesovsky &lt;kendy at collabora dot com&gt;
  > I confirm it is OK to use the Collabora name for the CODE NixOS
  > packages, provided that
  > 
  > * you keep the limits on the documents & connections, and
  > * you use the CODE branding for the builds

mmilata committed:
libreoffice: add collabora variant

collabora-online: init

mmilata committed:
libreoffice-online: init at 7.0.1.1

mmilata committed:
nixos/loolwsd: init

Add LibreOffice WebSocket Daemon.

mmahut commented:
Thank you for working on this, it would be very useful to have this in nixpkgs.

mmahut commented:
@GrahamcOfBorg build libreoffice-online

mmahut commented:
@GrahamcOfBorg build libreoffice-online

mmilata commented on pkgs/applications/office/libreoffice/libreoffice-srcs-additions.json:
```diff
@@ -1,5 +1,15 @@
   "ODFVALIDATOR": {"subdir": "../extern/"},
-  "OFFICEOTRON": {"subdir": "../extern/"}
+  "OFFICEOTRON": {"subdir": "../extern/"},
+  "FONT_NOTO_MORE": {"baseurl": "https://noto-website-2.storage.googleapis.com/pkgs"},
```
Hmm, this downloads 1.1G of fonts which are then not used because we build with `--without-fonts`. Perhaps implement `skip` instead of `baseurl` to save space & time?

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review-may-2019/3032/120

mmahut commented:
@GrahamcOfBorg build libreoffice-online collabora-office  collabora-online

mmilata commented on pkgs/applications/office/libreoffice/online/default.nix:
```diff
@@ -0,0 +1,114 @@
+  ];
+
+  passthru = {
+    libreoffice = libreoffice-core;
```
Probably have to add `libreoffice-core` to `propagatedBuildInputs`? I don't see it in `nix-store -qR /nix/store/xyz-libreofficeonline` and `ldd loolwsd` doesn't show it either.

mmahut commented:
@aanderse I know you are pretty busy, but would you be able to glance at this, please?

aanderse commented:
> @aanderse I know you are pretty busy, but would you be able to glance at this, please?

@mmahut the big problem for me to review this is that the application in this PR is **so** massive. To move forward with this PR I would suggest we break it down into fairly topic specific areas to review and find subject experts on each of those topics. I would also suggest we find someone with more than basic LibreOffice/Collabora Online experience to review this as well.

If a test could be written that would also be fantastic - even if that test can't be executed by the bot, it would give people the ability to setup a basic config, in a container perhaps, and run the test locally, poke around the internals, ensure everything is working as it should be.

I think this module would be of high value to the NixOS ecosystem, but without proper review and validation the best I could suggest is that maybe this make its way to NUR and gestate there... so shall break this PR down into conceptual components and find some subject experts to review individual parts?

mmilata commented:
@aanderse thanks for looking into this peculiar piece of software. I've added a simple test, though I can't get it to run on my system (even though in libvirt qemu almost identical configuration works fine), investigating. Anyway there's not much to test (without using something like Selenium) besides the service successfully starting.

Please let me know what to do to make this easier. I can provide access to a VM where this is deployed to anyone interested, just send me your ssh key over IRC (my nick is b42). If this turns out to be difficult I agree we can move this to NUR for the time being.

mmilata committed:
libreoffice: don't download unused fonts

Some of them are very large and they are not used because we pass
--without-fonts to configure.

7c6f434c commented:
For the record, we do have a VM test where a web browser is launched, then some keystrokes are sent to click the desired element, then OCR verifies if the expected text is on the screen. WebDriver would be indeed better in this case, but probably doesn't have an example to start with.

Mic92 commented:
> For the record, we do have a VM test where a web browser is launched, then some keystrokes are sent to click the desired element, then OCR verifies if the expected text is on the screen. WebDriver would be indeed better in this case, but probably doesn't have an example to start with.

That sounds quite brittle. I would prefer a simpler test.

Mic92 commented on pkgs/applications/office/libreoffice/online/kit-chroot-symlinks.patch:
```diff
@@ -0,0 +1,81 @@
+Author: Martin Milata <martin@martinmilata.cz>
+Date:   Thu Jan 30 17:44:31 2020 +0100
+
+    tdf#129895: handle symlinks when populating chroot
```
Did you also send this to upstream?

mmilata replied:
yup https://gerrit.libreoffice.org/c/online/+/87749

Mic92 commented on pkgs/applications/office/libreoffice/online/loolforkit-env-path.patch:
```diff
@@ -0,0 +1,16 @@
+ #else
+     std::string forKitPath = Path(Application::instance().commandPath()).parent().toString() + "loolforkit";
++
++    const char *envPath = std::getenv("LOOL_FORKIT_PATH");
```
Upstream might also accept this one.

Mic92 commented on pkgs/applications/office/libreoffice/online/node-packages.nix:
```diff
@@ -0,0 +1,4675 @@
+# This file has been generated by node2nix 1.7.0. Do not edit!
```
I would prefer this one to be integrated into `pkgs/development/node-packages/node-packages-v10.json`:

We do this for other packages as well, i.e.: 
```
, { "iosevka-build-deps": "../../data/fonts/iosevka" }
```

Thanks!

Mic92 commented:
One question regarding `libreoffice-online` and `collabora-office`: Which one do we need for nextcloud and which one we don't need?

Mic92 commented:
I would have a beefy machine to compile this beast.

mmilata commented:
> One question regarding `libreoffice-online` and `collabora-office`: Which one do we need for nextcloud and what is the functional difference.

For nextcloud you can choose between `libreoffice-online` (which has `libreoffice-fresh` as a dependency) or `collabora-online` (which has `collabora-office`) as a dependency. The `collabora-*` packages are based on older version of LibreOffice and are supposed to be more stable/polished. I think most nextcloud users use `collabora-online` because they have docker image available that is easy to use.

Mic92 commented:
> > One question regarding `libreoffice-online` and `collabora-office`: Which one do we need for nextcloud and what is the functional difference.
> 
> For nextcloud you can choose between `libreoffice-online` (which has `libreoffice-fresh` as a dependency) or `collabora-online` (which has `collabora-office`) as a dependency. The `collabora-*` packages are based on older version of LibreOffice and are supposed to be more stable/polished. I think most nextcloud users use `collabora-online` because they have docker image available that is easy to use.

Ok. Does `libreoffice-online` has any other use cases that might be relevant for nixpkgs? Otherwise I would go with `collobora-online`.

tobiasBora commented:
That's a very interesting PR ! Just, if `collobora-online` forces nixos to put a limit on the number of editors to avoid company use, would it make sense to support `libreoffice-online` if it does not have that constraint ?

mmilata commented:
@tobiasBora personally I'm not working on this anymore but I heard @sorki is interested in giving it a try.

Regarding the constraint it can be removed as long as we name the result something else than LibreOffice Online or Collabora Online. NixOS Office Online perhaps (NOOO)?:smile:

sorki commented:
Rebased and updated to `7.0.1.1`.

Changes:
- dropped nextcloud test as it requires internet connection for plugin download, will replace it with simple one testing that `loolwsd` responds to queries
- increased `systemplate` startup timeout
- switched proxy urls to `::1`
- squashed into 4 commits

Running this in a VM, needs bit more testing and polish but it looks good so far.

Pacman99 commented:
I was trying out this PR with the default settings. I imported the module and added libreoffice-online to pkgs with an overlay. Everything built alright, but I ran into an error with the systemd service:
```
Matching LibreOffice version not found, reinitializing systemplate, this may take a while ...
error: cannot open connection to remote store 'daemon': reading from file: Connection reset by peer
cp: missing destination file operand after '/var/lib/lool/systemplate/nix/store'
Try 'cp --help' for more information.
mv: cannot stat '/var/lib/lool/systemplate/nix/store/8bcc24znygvjgxxhn1hxlag753w112ck-libreoffice': No such file or directory
```
The error seems to be happening with the `nix-store` command in the systemplate setup script. But interestingly the script runs without error if I run it manually. So I think the nix daemon is unreachable by the script when run as a systemd service for some reason. I should also note that I am on nixUnstable.

Pacman99 commented on nixos/modules/services/web-apps/loolwsd/default.nix:
```diff
@@ -0,0 +1,245 @@
+  pkg = cfg.package;
+  corePkg = pkg.passthru.libreoffice;
+  subdir = corePkg.pname;
+  cmdLineOpts = {
```
```suggestion
  cmdLineOpts = {
    server_name = cfg.proxy.hostname;
```
This should be empty if there is no hostname, so the option should have a default. Otherwise it can be set if only proxy.hostname is set.

Pacman99 commented on nixos/modules/services/web-apps/loolwsd/default.nix:
```diff
@@ -0,0 +1,245 @@
+      enable = mkEnableOption "nginx proxy for loolwsd daemon";
+
+      hostname = mkOption {
+        type = str;
```
```suggestion
        type = str;
        default = "";
```
So the server_name can always be set

Pacman99 commented on nixos/modules/services/web-apps/loolwsd/default.nix:
```diff
@@ -0,0 +1,245 @@
+    # LO treats allowed_languages as a list of prefixes, by taking the first two letters we
+    # should be able to avoid mismatches of package name with the corresponding locale name
+    "allowed_languages" = concatMapStringsSep " " (builtins.substring 0 2) cfg.dictionaries;
+  } // cfg.settings;
```
server_name should be set if proxy.hostname is se

Pacman99 replied:
I sent a PR to implement this here:  otevrenamesta/nixpkgs#1

sorki committed:
fixup! nixos/loolwsd: init

sorki committed:
fixup! libreoffice-online: init at 7.0.1.1

sorki commented:
> But interestingly the script runs without error if I run it manually.

Is this possibly a race condition and your nix-daemon is restarting as well? 

To clarify - are you running a script from `ExecStart` manually or does it work when you start the service with `systemctl start`?

Thanks for looking into this @Pacman99! I've rebased this to make it easier and running a build to test it. I'll address your suggestions or you can submit a PR to `lool` branch of our fork if you wish.

Pacman99 commented:
> Is this possibly a race condition and your nix-daemon is restarting as well?
>
> To clarify - are you running a script from ExecStart manually or does it work when you start the service with systemctl start?

Both, it happens after a `nixos-rebuild switch` and when I run systemctl start loolwsd. And yeah, I ran the script from ExecStartPre manually and thats when it worked. Which is an ok workaround since it only has to be run on libreoffice updates, but its not optimal.

>Thanks for looking into this @Pacman99! I've rebased this to make it easier and running a build to test it. I'll address your suggestions or you can submit a PR to lool branch of our fork if you wish.

Thanks to both of you for working on this module! Yeah I can probably submit a PR soon to add the features I mentioned.

Commit:
nixos/loolwsd: set server_name setting when proxy is enabled

Commit:
nixos/loolwsd: add option for setting loolwsd trusted host

sorki committed:
nixosTests.loolwsd: init

sorki commented:
Added a simplified test without nextcloud. Sadly `loolforkit` now dumps core - https://gist.github.com/sorki/48997cb80f85b9f4bfde8dbc0e39e6fb which seems to be caused by the rebase. I've pushed a prior-rebase backup to https://github.com/otevrenamesta/nixpkgs/tree/lool-b to be able to bisect.

Pacman99 commented:
I narrowed down my error further. The lool user cannot connect to the nix daemon to use `nix-store`. I was able to get the same error like this:
```
sudo -u lool bash
nix-store --query --requisites /nix/store/8bcc24znygvjgxxhn1hxlag753w112ck-libreoffice-7.0.4.2/
```
which outputs:
```
error: --- Error ----------------------------------------------------- nix-store
cannot open connection to remote store 'daemon': error: --- SysError -------------------------------------------------- nix-store
reading from file: Connection reset by peer
```

sorki commented:
Maybe the chroot closure can be prepared before, so the `nix-store --query` call is not needed in service.

I was tracking down the SIGABRT issue and the last relevant strace message is the `loolforkit` service loading
`/nix/store/gnrc4wjydnvz74mh6z683yd63k59sh7g-dejavu-fonts-2.37/share/fonts/truetype/DejaVuMathTeXGyre.ttf`. Pretty weird.

Pacman99 commented:
I figured out my issue, its because I had `nix.allowedUsers` set, so the lool user wasn't able to connect to the nix daemon.

I don't think this is something to worry about since `allowedUsers` is set to "*" by default and some other nixos modules work on this assumption.

sorki commented:
> I figured out my issue, its because I had `nix.allowedUsers` set, so the lool user wasn't able to connect to the nix daemon.
> 
> I don't think this is something to worry about since `allowedUsers` is set to "*" by default and some other nixos modules work on this assumption.

Sounds good.

Btw - you have this based on older commit or using this rebased version? If it's on the older one would you upload it somewhere or just paste the nixpkgs commit it is based on so I can possibly bisect between that and the commit this is currently based on?

Pacman99 commented:
I believe I'm using an older commit:
`a1cbe120c84d6b8125f889cef3713e2dd1762c14`
But I'm not entirely sure because I just copied the files to my server's flake.

This is the module I'm using: https://gitlab.com/coffeetablebrothers/myrdd/-/blob/master/modules/loolwsd.nix
And in the same repo I coped over the libreoffice-online package into the `pkgs` folder.

Pacman99 commented on pkgs/applications/office/libreoffice/default.nix:
```diff
@@ -1,4 +1,4 @@
-{ stdenv, fetchurl, lib, pam, python3, libxslt, perl, ArchiveZip, box2d, gettext
+{ stdenv, fetchurl, fetchFromGitHub, lib, pam, python3, libxslt, perl, ArchiveZip, box2d, gettext
```
```suggestion
{ stdenv, fetchurl, fetchFromGitHub, fetchpatch, lib, pam, python3, libxslt, perl, ArchiveZip, box2d, gettext
```
needed for collabora

drupol commented:
What's the status of this ?

RaitoBezarius commented:
How can I help to get this PR merged? It is super cool!

mmilata commented:
AFAIK nobody is working on this. To get it merged you'd need to take over the branch, update to latest nixpkgs & loolwsd, and address the outstanding issues ...

drupol commented:
@RaitoBezarius You think need to talk with @Minion3665

Minion3665 commented:
Thanks the the ping pol,

@RaitoBezarius I started looking at this in the summer but haven't worked on it for a while due to other responsibilities

I'd be happy to work with you on it now though if you're interested

RaitoBezarius commented:
> Thanks the the ping pol,
> 
> @RaitoBezarius I started looking at this in the summer but haven't worked on it for a while due to other responsibilities
> 
> I'd be happy to work with you on it now though if you're interested

I might not have full bandwidth to drive the efforts by myself, but I am very interested into following the work and helping as much as I can. It would be really nice to have this in NixOS.

Mic92 commented:
Closing because of lack of activity. Please re-open if you start working on this again.
